### PR TITLE
Add batch withdraw to roster view

### PR DESCRIPTION
**Pull request by lukenewman**
## Background
This PR adds the final batch action, Withdraw, to the roster redesign.

## GitHub Issue
https://github.com/ctoec/data-collection/issues/1364

## Associated PRs
[This PR](https://github.com/ctoec/data-collection/pull/1562) may need to be merged in before this one. Once withdrawn, records are still showing up in the roster but without any enrollment data.

## Validation Plan
* [ ] - on the roster page, select one record and click Withdraw, fill out the form, hit Withdraw, record should be withdrawn
* [ ] - on the roster page, select multiple record and click Withdraw, fill out the form, hit Withdraw, records should be withdrawn

## Tech Specs
- [ ] If any TypeORM entities were created/updated/removed, the necessary migration(s) were [correctly auto-generated](https://github.com/typeorm/typeorm/blob/master/docs/using-cli.md#generate-a-migration-from-existing-table-schema) and **not manually written**.

## Automated Testing
- [ ] All unit tests are passing.
- [ ] Unit tests have been added to cover this changeset.

## Additional Context
N/A

### CODE REVIEW

**Comment by lukenewman** [comment]
merged in jesse's recent enrollment pr, and the old problem has been fixed (children showing up in roster without enrollment data). however, now when selecting multiple children, not all of them are being withdrawn. investigating...

**Review by nickbristow — APPROVED** [review]
LGTM 

Edit: just need to fix unused var alerts

**Comment by lukenewman** [comment]
this is good to go

File: client/src/containers/Roster/BatchDelete.tsx [hunk 1]
```diff
@@ -96,8 +96,8 @@ export const BatchDelete: React.FC<DeleteProps> = ({
         header={
           <h2>{`Delete enrollment for ${
             isOneChild
-              ? Object.values(selectedChildren)[0]
-              : Object.keys(selectedChildren).length + ' records'
+              ? Object.values(children)[0]
+              : Object.keys(children).length + ' records'
```
Comment (jessethach) [inline client/src/containers/Roster/BatchDelete.tsx:100]: Nit: Any reason to not place in template string like the others?

File: client/src/containers/Roster/BatchWithdraw.tsx [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        content={
+          <>
+            <Form<WithdrawRequest>
+              onSubmit={onSubmit}
+              data={{} as WithdrawRequest}
+              className="usa-form"
+            >
+              <ExitReasonField<WithdrawRequest> />
+              <EnrollmentEndDateField<WithdrawRequest> />
+              {/* does this need to still be here? */}
```
Comment (jessethach) [inline client/src/containers/Roster/BatchWithdraw.tsx:122]: Ever find out if this needs to be here?

File: client/src/containers/Roster/Roster.tsx [hunk 3]
```diff
@@ -374,17 +371,31 @@ const Roster: React.FC = () => {
                     setAlerts={setAlerts}
                     fetchRoster={fetchRoster}
                   />
-                  <TableBatchAction
+                  <BatchWithdraw
+                    children={(() => {
+                      let c: ChildEnrollmentIdToNameMap = {};
```
Comment (jessethach) [inline client/src/containers/Roster/Roster.tsx:376]: Might be a good use case for `.reduce`:

```
return selectedRows.reduce<ChildEnrollmentIdToNameMap>((acc, { id: rowId }) => {
  const child = childRows.find(
    (child) => child.id === rowId
  );
  if (child?.enrollmentId) {
    acc[child.enrollmentId] = child?.name;
  }
  return acc;
}, {});
```

File: src/routes/enrollments.ts [hunk 4]
```diff
@@ -62,12 +62,24 @@ enrollmentsRouter.post(
...
   passAsyncError(async (req, res) => {
-    const enrollmentId = parseInt(req.params['enrollmentId']);
+    const query = req.query['enrollmentIds'] as string[];
+    const enrollmentIds = query.map(Number);
+    if (enrollmentIds === null || enrollmentIds.length === 0) {
+      throw new HttpErrors.BadRequest(
+        'null or empty enrollmentIds array provided.'
+      );
+    }
+    console.log('enrollmentIds', enrollmentIds);
```
Comment (jessethach) [inline src/routes/enrollments.ts:80]: Leftover log?

**Review by jessethach — APPROVED** [review]
Just optional nits. Great work!

File: client/src/containers/Roster/BatchWithdraw.tsx (hunk 2 above)
Comment (lukenewman) [reply client/src/containers/Roster/BatchWithdraw.tsx:122]: haha -- nope @JordanGuinn do we need to keep that field around?

File: client/src/containers/Roster/BatchDelete.tsx (hunk 1 above)
Comment (lukenewman) [reply client/src/containers/Roster/BatchDelete.tsx:100]: like what others?

Comment (jessethach) [reply client/src/containers/Roster/BatchDelete.tsx:100]: Like line 75 and 76 above

Comment (lukenewman) [reply client/src/containers/Roster/BatchDelete.tsx:100]: ah. i mean, it's basically two ways of doing the same thing. they both use templates...

**Review by bamader — APPROVED** [review]

**Comment by bamader** [comment]
## Validation Plan

- [ ] Can batch withdraw single child
- [ ] Can successfully batch withdraw multiple children with no failures
- [ ] Can batch withdraw children and have some of them fail

**Comment by bamader** [comment]
Validations come back green on this PR, but I did find a pair of other things that could be worth attention as bugs.

Bug 1: for some reason, using the batch actions does not seem to clear alerts from the roster page, so they wind up accumulating at the top (even with separate/different actions like uploads, successful withdraws, failed withdraws, etc.). See image:
Screen Shot 2021-11-17 at 2 40 13 PM
 
Bug 2: After withdrawing some children, the counts displayed on the roster don't update to reflect withdrawn numbers. You can see that after withdrawals, the number of enrollments go down (because children with active enrollments decreased), but we still report this as the total number of children and give age-group breakdowns that don't account for withdrawn records.
Screen Shot 2021-11-17 at 2 40 28 PM

File: src/routes/enrollments.ts (hunk 4 above)
Comment (JordanGuinn) [reply src/routes/enrollments.ts]: Reminder to pull this out

**Review by JordanGuinn — APPROVED** [review]

**Comment by JordanGuinn** [comment]
That first bug seems like it's own thing, and should be addressed on it's own.  Wondering about the second bug though - does this happen with any other batch actions?  If so, then I'd say it should _also_ be it's own bug.

@bamader @lukenewman

**Comment by bamader** [comment]
@JordanGuinn I'll give bug 2 a test with batch delete to see if it correctly handles counts (which I'd imagine it does). I think we've got a batch change enrollment too so I'll test that. But yeah first bug is probs its own thing.

**Comment by bamader** [comment]
Looks like batch delete still counts the kids you just deleted until the app is refreshed. Batch change age group seems unaffected by this (outside of problems displaying historical enrollments, but that's a separate PR).

**Comment by lukenewman** [comment]
@bamader was there a conclusion reached as to whether this is good to go in? i'm not seeing batch delete still counting the kids you just deleted. also not seeing that on this batch withdraw work either. however, the alerts are persisting pretty far. i can get rid of the `...current` to most likely fix that

**Comment by bamader** [comment]
@lukenewman Hmm odd that you can't repro the bug. Let me give it another try. The only thing I was seeing with batch delete was that you had to refresh the page to get the app to update the counts. The alerts being persistent I think we decided should be a separate bug ticket rather than a blocker to this, so I'd say either you or Jordan can write a ticket for it and then once we clear up the counts thing we'll be ready.

**Comment by bamader** [comment]
Screen Shot 2021-11-19 at 1 23 47 PM

Okay, so it looks like the counting issue is only happening in the header section. The number in the roster section is correct, but the numbers up top (13 children across 8 sites) is wrong, as are the counts of enrollments in each age group (because it doesn't include those who were withdrawn).

I think this was also a problem in some stuff Nick was working on, and I believe he said it was happening because the endpoint was retrieving too many rows and counting all of them rather than filtering out the withdrawn children. @nickbristow Have you already addressed this in your PR, or is it something Lucas needs to work on here?

**Comment by lukenewman** [comment]
@bamader that child count bug should be fixed

**Comment by GordonFarrell** [comment]
Child count fix is working properly, batch withdrawal happy path also working properly.

I am seeing a bit of a bug though. If a batch withdrawal is attempted but fails due to not having selected a reason or dates, even after adding the missing data it will still give an "Exit date and exit reason are required" bad request failure unless you completely close the modal then reopen and try again.

**Comment by lukenewman** [comment]
@GordonFarrell can you expand on those repro steps? couldn't get that bug to repro

**Comment by GordonFarrell** [comment]
Sure thing!

1. Select several students and open withdrawal modal

2. Leave reason empty, fill two valid dates

3. Click withdraw and get an error

4.Add a reason and click withdraw again

For me the error is persisting until I close the modal entirely and try again from the start with all valid info for each field.

**Comment by bamader** [comment]
If I'm not mistaken that bug exists in the normal withdraw functionality, not introduced by this PR's usage of the withdraw modal. We can definitely fix it here (because it makes sense to finalize withdraw functionality), but if we want to get this shipped sooner rather than later we can open a separate ticket for that.

**Comment by bamader** [comment]
Confirmed, the exit reason bug exists independent of this PR. @JordanGuinn What should be the call here? Fix it up as part of this PR, or create its own bug ticket (in the interest of giving us closed tickets to report this week)?

**Comment by JordanGuinn** [comment]
@bamader If it's not a super quick fix, let's create a separate ticket and track it independently

**Comment by bamader** [comment]
Got it. I don't think it will be a quick fix (it's likely a 2 or 4 pointer, just guessing) so I'm in favor of separate tracking. Once you create that ticket, this PR will be good to go!

**Comment by JordanGuinn** [comment]
Ticket created - https://app.zenhub.com/workspaces/ece-dev-board-5ff506028d35f30012e0e937/issues/ctoec/data-collection/1480

**Pull request merged**